Update VM based installation for 4.3 VM image (#2716)
uyuni-project/uyuni-docs: modules/installation-and-upgrade/pages/install-vm.adoc

If you want to use a production environment, review the requirements listed in xref:installation-and-upgrade:hardware-requirements.adoc[].
====

[cols="1,1", options="header"]
|===

> 2+<| Virtual Machine Settings overview

Reviewer: ```suggestion
2+<| Virtual machine settings overview
```

| Installation Method | Import Existing Disk Image
| OS:                 | SUSE Linux Enterprise 15 SP4
| Memory:             | 16 GB
| CPU's:              | 4
| Virtual Disks:      |
| VirtIO Disk 1       | SUSE-Manager-Server.x86_64-4.3.10-KVM.qcow2
| VirtIO Disk 2       | 101 GB for [path]``/var/spacewalk``
| VirtIO Disk 3       | 50 GB for [path]``/var/lib/pgsql``

> | VirtIO Disk 4       | 4 GB for swap

Reviewer 2: With adding the disk as swap here there is the problem down in the `suma-storage` command that this swap partition is not being used, we haven't mounted it properly. Is there a way to add it in the `suma-storage`? Or another way to use it in the VM? Otherwise the SUMA configuration creates a 2GB swap from the root partition on its own.

Author: Good point. I missed the swap entry. Size of the root is set to 15GB, when image is deployed, there are 11GB available. So taking 2GB from it is kind of big chunk of available storage.

`suma-storage` is maintained by cloud team at https://github.com/SUSE-Enceladus/susemanager-cloud-setup so it will need to be discussed with them or create PR for it there.

Reviewer 2: Ok great then if we can create a new PR or a card for that, it would be a separate issue and we can resolve this. I will resolve the conversation here to be able to merge this current PR, it should go out as soon as possible with the next update.

For more information on {sle} Virtualization Guide, see https://documentation.suse.com/sles/15-SP4/html/SLES-all/book-virtualization.html.
====

[IMPORTANT]
====

> {productname} VM image does not setup [systemitem]``root`` or any other user account. User or [systemitem]``root`` authentications need to be setup during first boot. This can be done using [systemitem]``Ignition`` or [systemitem]``Cloud-Init`` methods.

Reviewer: ```suggestion
{productname} VM image does not setup [systemitem]``root`` or any other user account.
User or [systemitem]``root`` authentications need to be setup during first boot.
This can be done using [systemitem]``Ignition`` or [systemitem]``Cloud-Init`` methods.
```

====

[[sumavm.ignition]]
== {productname} basic configuration using [systemitem]``Ignition``

[systemitem]``Ignition`` is a provisioning tool that enables you to configure a system according to your specification on the first boot.
When the system is booted for the first time, [systemitem]``Ignition`` is loaded as part of an initramfs and searches for a configuration file within a specific directory (on a USB flash disk, or you can provide a URL).

> [systemitem]``Ignition`` uses a configuration file in the JSON format. The file is called [path]``config.ign``.

Reviewer: ```suggestion
[systemitem]``Ignition`` uses a configuration file in the JSON format.
The file is called [path]``config.ign``.
```

> The [path]``config.ign`` is a JSON configuration file that provides prescriptions for Ignition. You can either create the file manually in JSON, or you can use the Fuel Ignition tool (see https://ignite.opensuse.org/) to generate a basic set of prescriptions. Bear in mind that the Fuel Ignition tool does not provide a full set of options, so you might have to modify the file manually.

Reviewer: ```suggestion
The [path]``config.ign`` is a JSON configuration file that provides prescriptions for Ignition.
You can either create the file manually in JSON, or you can use the Fuel Ignition tool to generate a basic set of prescriptions.
The Fuel Ignition tool does not provide a full set of options, so you might have to modify the file manually.
For more information, see https://ignite.opensuse.org/. 
```

When installing, the configuration file [path]``config.ign`` must reside in the [path]``ignition`` subdirectory on the configuration media labeled [systemitem]``ignition``. The directory structure must look as follows:

```
<root directory>
└── ignition
    └── config.ign
```

If you intend to configure a QEMU/KVM virtual machine, provide the path to the [path]``config.ign`` as an attribute of the qemu command. For example:

```
-fw_cfg name=opt/com.coreos/config,file=PATH_TO_config.ign
```

> The config.ign contains various data types: objects, strings, integers, Booleans and lists of objects. For a complete specification, refer to Ignition specification v3.3.0 (https://coreos.github.io/ignition/configuration-v3_3/).

Reviewer: ```suggestion
The config.ign contains various data types: objects, strings, integers, booleans and lists of objects.
For the complete specification, see https://coreos.github.io/ignition/configuration-v3_3/.
```

=== Setup root password using Ignition

The [systemitem]``passwd`` attribute is used to add users. If you intend to log in to your system, create root and set the root's password and/or add the SSH key to the [systemitem]``Ignition`` configuration. You need to hash the root password, for example, by using the openssl command:

> ```
> openssl passwd -6
> ```

Reviewer: ```suggestion
----
openssl passwd -6
----
```

The command creates a hash of the password you chose. Use this hash as the value of the [systemitem]``password_hash`` attribute.

The users attribute must contain at least one [systemitem]``name`` attribute. [systemitem]``ssh_authorized_keys`` is a list of ssh keys for the user. 

Prepare required configuration using following commands:

```
$ mkdir -p root/ignition

> $ cat << EOF > root/ignition/config.ign

Reviewer 2: This command mangles the hash in the terminal and in the end you end up with something completely different which is invalid. 
```suggestion
Create a file named root/ignition/config.ign with the following content:
```

{
  "ignition": {
    "version": "3.2.0"
  },
  "passwd": {
    "users": [
      {
        "name": "root",
        "passwordHash": "$2a$10$qV298UV11u9lCFDjpHpCUe1cErBiVR.G3shukxs3.2PAO1xhJWs0K"
      }
    ]
  }
}

> EOF

Reviewer 2: ```suggestion
```

> ```

Reviewer: ```suggestion
----
mkdir -p root/ignition
cat << EOF > root/ignition/config.ign
{
  "ignition": {
    "version": "3.2.0"
  },
  "passwd": {
    "users": [
      {
        "name": "root",
        "passwordHash": "$2a$10$qV298UV11u9lCFDjpHpCUe1cErBiVR.G3shukxs3.2PAO1xhJWs0K"
      }
    ]
  }
}

EOF
----

> Prepare [systemitem]``Ingition`` ISO file using the command:

Reviewer 2: ```suggestion
Prepare the [systemitem]``Ignition`` ISO file using the command:
```

```
mkisofs -full-iso9660-filenames -o suma_ignition.iso -V ignition root
```

> Attach created [path]``suma_ignition.iso`` file as a volume to the creating virtual machine. This particular example is setting [systemitem]``root`` password to `linux`.

Reviewer 2: ```suggestion
Attach the created [path]``suma_ignition.iso`` file as a volume to the creating virtual machine. This particular example is setting [systemitem]``root`` password to `linux`.
```

> For more information about [systemitem]``Ignition``, see https://documentation.suse.com/sle-micro/5.4/single-html/SLE-Micro-deployment/#cha-images-ignition

Reviewer: ```suggestion
For more information about [systemitem]``Ignition``, see https://documentation.suse.com/sle-micro/5.4/single-html/SLE-Micro-deployment/#cha-images-ignition.
```

[[sumavm.cloud_init]]
== {productname} basic configuration using Cloud Init disk

[systemitem]``Cloud Init`` is a provisioning tool that enables you to configure a system according to your specification on the first boot.
When the system is booted for the first time, [systemitem]``Cloud Init service`` is loaded and searches for a configuration file within a specific directory (on a USB flash disk, or you can provide a URL).

[systemitem]``Cloud Init`` uses few configuration files in the YAML format. Used files are named [path]``meta-data``, [path]``network-config`` and [path]``user-data``.

[systemitem]``Cloud Init`` allows numerous sources where to store configuration data. In this guide we use local iso image with volume id [systemitem]``cidata`` as a source. The directory structure must look as follows:

> ```
> <root directory>
> └── meta-data
> └── network-config
> └── user-data
> ```

Reviewer: ```suggestion
----
<root directory>
└── meta-data
└── network-config
└── user-data
----
```

If you intend to configure a QEMU/KVM virtual machine, provide the path to the [path]``config.ign`` as an attribute of the qemu command. For example:

```
-fw_cfg name=opt/com.coreos/config,file=PATH_TO_config.ign
```

The [systemitem]``Cloud Init`` allows many management options. For a complete specification, refer to Cloud Init specification (https://cloudinit.readthedocs.io/en/latest/index.html).

=== Setup root password using Cloud Init

You need to hash the root password, for example, by using the openssl command:

> ```
> openssl passwd -6
> ```

Reviewer: ```suggestion
----
openssl passwd -6
----
```

The command creates a hash of the password you chose. Use this hash as the value of the [systemitem]``password`` attribute.

> Prepare needed configuration files using following commands:

Reviewer 2: ```suggestion
Prepare the needed configuration files using the following commands:
```

```
$ touch network-config
$ touch meta-data

> $ cat << EOF > user-data

Reviewer 2: ```suggestion
Create a file named user-data with the following content:
```
this is the same issue as ignition, the terminal mangles the hash, you need to escape special characters etc

#cloud-config
chpasswd:
  expire: false
  users:
    - name: root
      password: $2a$10$qV298UV11u9lCFDjpHpCUe1cErBiVR.G3shukxs3.2PAO1xhJWs0K

> EOF

Reviewer 2: ```suggestion
```

> ```
> $ touch network-config
> $ touch meta-data
> $ cat << EOF > user-data
> #cloud-config
> chpasswd:
>   expire: false
>   users:
>     - name: root
>       password: $2a$10$qV298UV11u9lCFDjpHpCUe1cErBiVR.G3shukxs3.2PAO1xhJWs0K
> 
> EOF
> ```

Reviewer: ```suggestion

----
touch network-config
touch meta-data
cat << EOF > user-data
#cloud-config
chpasswd:
  expire: false
  users:
    - name: root
      password: $2a$10$qV298UV11u9lCFDjpHpCUe1cErBiVR.G3shukxs3.2PAO1xhJWs0K

EOF
----
```

Prepare [systemitem]``Cloud Init`` ISO file using the command:

> ```
> mkisofs -rational-rock -joliet -o suma_cloudinit.iso -V cidata network-config meta-data user-data
> ```

Reviewer: ```suggestion

----
mkisofs -rational-rock -joliet -o suma_cloudinit.iso -V cidata network-config meta-data user-data
----
```

> Attach created [path]``suma_cloudinit.iso`` file as a volume to the creating virtual machine. This particular example is setting [systemitem]``root`` password to `linux`.

Reviewer 2: ```suggestion
Attach the created [path]``suma_cloudinit.iso`` file as a volume to the creating virtual machine. This particular example is setting [systemitem]``root`` password to `linux`.
```

[[sumavm.kvm.settings]]
== {productname} Virtual Machine Settings

Create three additional virtual disks required for the {productname} storage partitions.

.Procedure: Creating the Required Partitions with KVM
. Create a new virtual machine using the downloaded {productname} KVM image and select [guimenu]``Import existing disk image``.

> . Set [literal]``SUSE Linux Enterprise 15 SP4`` as installed operating system.

Reviewer 2: ```suggestion
. Set [literal]``SUSE Linux Enterprise 15 SP4`` as the installed operating system.
```

For more information on {sle} Virtualization Guide, see https://documentation.suse.com/sles/15-SP4/html/SLES-all/book-virtualization.html.
====

[IMPORTANT]
====

> {productname} VM image does not setup [systemitem]``root`` or any other user account.
> User or [systemitem]``root`` authentications need to be setup during first boot.

Reviewer: ```suggestion
{productname} VM image does not set up [systemitem]``root`` or any other user account.
User or [systemitem]``root`` authentications need to be set up during first boot.
```

====

[[sumavm.ignition]]
== {productname} basic configuration using [systemitem]``Ignition``

[systemitem]``Ignition`` is a provisioning tool that enables you to configure a system according to your specification on the first boot.
When the system is booted for the first time, [systemitem]``Ignition`` is loaded as part of an initramfs and searches for a configuration file within a specific directory (on a USB flash disk, or you can provide a URL).

[systemitem]``Ignition`` uses a configuration file in the JSON format.
The file is called [path]``config.ign``.

The [path]``config.ign`` is a JSON configuration file that provides prescriptions for Ignition.
You can either create the file manually in JSON, or you can use the Fuel Ignition tool to generate a basic set of prescriptions.
The Fuel Ignition tool does not provide a full set of options, so you might have to modify the file manually.
For more information, see https://ignite.opensuse.org/. 

When installing, the configuration file [path]``config.ign`` must reside in the [path]``ignition`` subdirectory on the configuration media labeled [systemitem]``ignition``. The directory structure must look as follows:

> ```
> <root directory>
> └── ignition
>     └── config.ign
> ```

Reviewer: ```suggestion
----
<root directory>
└── ignition
    └── config.ign
----
```

> If you intend to configure a QEMU/KVM virtual machine, provide the path to the [path]``config.ign`` as an attribute of the qemu command. For example:

Reviewer: ```suggestion
If you intend to configure a QEMU/KVM virtual machine, provide the path to the [path]``config.ign`` file as an attribute of the qemu command. For example:
```

```
-fw_cfg name=opt/com.coreos/config,file=PATH_TO_config.ign
```

> The config.ign contains various data types: objects, strings, integers, booleans and lists of objects.

Reviewer: ```suggestion
The [path]``config.ign`` file contains various data types: objects, strings, integers, booleans, and lists of objects.
```

For the complete specification, see https://coreos.github.io/ignition/configuration-v3_3/.

> === Setup root password using Ignition

Reviewer: ```suggestion
=== Set up root password using Ignition
```

The [systemitem]``passwd`` attribute is used to add users. If you intend to log in to your system, create root and set the root's password and/or add the SSH key to the [systemitem]``Ignition`` configuration. You need to hash the root password, for example, by using the openssl command:

----
openssl passwd -6
----

The command creates a hash of the password you chose. Use this hash as the value of the [systemitem]``password_hash`` attribute.

The users attribute must contain at least one [systemitem]``name`` attribute. [systemitem]``ssh_authorized_keys`` is a list of ssh keys for the user. 

Prepare required configuration using following commands:

----
mkdir -p root/ignition
cat << EOF > root/ignition/config.ign
{
  "ignition": {
    "version": "3.2.0"
  },
  "passwd": {
    "users": [
      {
        "name": "root",
        "passwordHash": "$2a$10$qV298UV11u9lCFDjpHpCUe1cErBiVR.G3shukxs3.2PAO1xhJWs0K"
      }
    ]
  }
}

----

Prepare the [systemitem]``Ignition`` ISO file using the command:

> ```
> mkisofs -full-iso9660-filenames -o suma_ignition.iso -V ignition root
> ```

Reviewer: ```suggestion

----
mkisofs -full-iso9660-filenames -o suma_ignition.iso -V ignition root
----
```

[[sumavm.cloud_init]]
== {productname} basic configuration using Cloud Init disk

[systemitem]``Cloud Init`` is a provisioning tool that enables you to configure a system according to your specification on the first boot.
When the system is booted for the first time, [systemitem]``Cloud Init service`` is loaded and searches for a configuration file within a specific directory (on a USB flash disk, or you can provide a URL).

[systemitem]``Cloud Init`` uses few configuration files in the YAML format. Used files are named [path]``meta-data``, [path]``network-config`` and [path]``user-data``.

[systemitem]``Cloud Init`` allows numerous sources where to store configuration data. In this guide we use local iso image with volume id [systemitem]``cidata`` as a source. The directory structure must look as follows:

----
<root directory>
└── meta-data
└── network-config
└── user-data
----

If you intend to configure a QEMU/KVM virtual machine, provide the path to the [path]``config.ign`` as an attribute of the qemu command. For example:

> ```
> -fw_cfg name=opt/com.coreos/config,file=PATH_TO_config.ign
> ```

Reviewer: ```suggestion
----
-fw_cfg name=opt/com.coreos/config,file=PATH_TO_config.ign
----
```

The [systemitem]``Cloud Init`` allows many management options. For a complete specification, refer to Cloud Init specification (https://cloudinit.readthedocs.io/en/latest/index.html).

> === Setup root password using Cloud Init

Reviewer: ```suggestion
=== Set up root password using Cloud Init
```

You need to hash the root password, for example, by using the openssl command:

----
openssl passwd -6
----

The command creates a hash of the password you chose. Use this hash as the value of the [systemitem]``password`` attribute.

Prepare the needed configuration files using the following commands:

> ```
> $ touch network-config
> $ touch meta-data
> Create a file named user-data with the following content:
> #cloud-config
> chpasswd:
>   expire: false
>   users:
>     - name: root
>       password: $2a$10$qV298UV11u9lCFDjpHpCUe1cErBiVR.G3shukxs3.2PAO1xhJWs0K
> 
> ```

Reviewer: ```suggestion

----
touch network-config
touch meta-data
----

Create a file named user-data with the following content:

----
#cloud-config
chpasswd:
  expire: false
  users:
    - name: root
      password: $2a$10$qV298UV11u9lCFDjpHpCUe1cErBiVR.G3shukxs3.2PAO1xhJWs0K
----
```

[cols="1,1,1", options="header"]
|===
| VirtIO Storage Disks | Name      | Sizing

> | VirtIO Disk 2        | spacewalk | 500{nbsp}GB

Reviewer 2: This has different values than above in the other table. We need to sync them and maybe consolidate them in one table. I would suggest to consolidate here after the conf files are created and while you are creating the VM settings.

| VirtIO Disk 3        | pgsql     | 100{nbsp}GB
| VirtIO Disk 4        | swap      | 4{nbsp}GB
|===

> . Click btn:[Add Hardware] to attach virual CDROM device with prepared [systemitem]``Ignition`` or [systemitem]``Cloud Init`` disk.

Reviewer 2: ```suggestion
. Click btn:[Add Hardware] to attach a virtual CDROM device with the prepared [systemitem]``Ignition`` or [systemitem]``Cloud Init`` disk.
```

> . Click btn:[Begin Installation] to boot the new VM from the {productname} image. Wait until login prompt is presented. Login using credentials set by configuration disk.

Reviewer 2: ```suggestion
. Click btn:[Begin Installation] to boot the new VM from the {productname} image. Wait until the login prompt is presented. Login using credentials set by configuration disk.
```

// Follow the prompts to complete the basic {minimalsles} installation, until the process is complete and the command prompt waits for input.

// During the basic installation prompts you are asked to enter the root password.
// In the next message box click btn:[Confirm root Password].

[[minimmal.susemgr.prep]]
== Preparing {productname} for SUSE Manager

[[proc.minimmal.susemgr.prep]]

> .Procedure: Preparing for {productname} Run

Reviewer: ```suggestion
.Procedure: Preparing for {productname} run
```

SUSEConnect -p sle-module-web-scripting/15.4/x86_64

> SUSEConnect -p sle-module-suse-manager-server/<productnumber>/x86_64

Reviewer 2: This command is already done above in lines 232. So maybe we remove this last one as well. 
```suggestion
```

Reviewer: I think someone has already removed the previous occurrence.  At least,I cannot find it.

----

. Prepare {productname} storage:
  [path]``suma-storage`` command automatically prepares and configures previously created external storage for use with {productname}.

>   In the following command first parameter is the device for {productname} data, second parameter is the device for the database.

Reviewer 2: ```suggestion
  In the following command the first parameter is the device for {productname} data, the second parameter is the device for the database.
```

+

----
suma-storage /dev/vdb /dev/vdc
----

> . Virtual machine is now ready for {productname} to be setup.

Reviewer 2: ```suggestion
. The virtual machine is now ready for {productname} to be set up.
```
